zkevm-circuits/src/taiko_pi_circuit.rs

```diff
@@ -306,6 +685,488 @@ impl<F: Field> SubCircuitConfig<F> for TaikoPiCircuitConfig<F> {
+                ),
+                (
+                    q_blk_hdr_rlp_end.expr() * blk_hdr_hash_hi,
+                    meta.query_advice(keccak_table2.output_hi, Rotation::cur()),
+                ),
+                (
+                    q_blk_hdr_rlp_end * blk_hdr_hash_lo,
+                    meta.query_advice(keccak_table2.output_lo, Rotation::cur()),
+                ),
+            ]
+        });
+
+        meta.lookup_any(
+            "Block header: Check hi parts of block hashes against previous hashes",
+            |meta| {
+                let q_blk_hdr_rlp_end = meta.query_selector(q_blk_hdr_rlp_end);
+                let blk_hdr_hash_hi = meta.query_advice(blk_hdr_blockhash, Rotation::cur());
+                let q_lookup_blockhash = meta.query_selector(q_lookup_blockhash);
+                let tag = meta.query_fixed(block_table_tag_blockhash, Rotation::prev());
+                let index = meta.query_fixed(block_table_index_blockhash, Rotation::cur());
+                let q_sel = and::expr([q_blk_hdr_rlp_end, q_lookup_blockhash]);
+                vec![
+                    (
+                        q_sel.expr() * tag,
+                        meta.query_advice(block_table_blockhash.tag, Rotation::cur()),
+                    ),
+                    (
+                        q_sel.expr() * index,
+                        meta.query_advice(block_table_blockhash.index, Rotation::cur()),
+                    ),
+                    (
+                        q_sel.expr() * blk_hdr_hash_hi,
+                        meta.query_advice(block_table_blockhash.value, Rotation::cur()),
+                    ),
+                ]
+            },
+        );
+        meta.lookup_any(
+            "Block header: Check lo parts of block hashes against previous hashes",
+            |meta| {
+                let q_blk_hdr_rlp_end = meta.query_selector(q_blk_hdr_rlp_end);
+                let blk_hdr_hash_lo = meta.query_advice(blk_hdr_blockhash, Rotation::prev());
+                let q_lookup_blockhash = meta.query_selector(q_lookup_blockhash);
+                let tag = meta.query_fixed(block_table_tag_blockhash, Rotation(-2));
+                let index = meta.query_fixed(block_table_index_blockhash, Rotation::cur());
+                let q_sel = and::expr([q_blk_hdr_rlp_end, q_lookup_blockhash]);
+                vec![
+                    (
+                        q_sel.expr() * tag,
+                        meta.query_advice(block_table_blockhash.tag, Rotation::cur()),
+                    ),
+                    (
+                        q_sel.expr() * index,
+                        meta.query_advice(block_table_blockhash.index, Rotation::cur()),
+                    ),
+                    (
+                        q_sel.expr() * blk_hdr_hash_lo,
+                        meta.query_advice(block_table_blockhash.value, Rotation::cur()),
+                    ),
+                ]
+            },
+        );
```

Comment:
Would also use a loop here to handle the lo/hi values because there's quite a lot of the same code that can be reused inside the loop.